File: city-service/src/main/java/io/github/xpakx/habitcity/shop/ShopService.java
package io.github.xpakx.habitcity.shop;

import io.github.xpakx.habitcity.equipment.dto.AccountEvent;
import io.github.xpakx.habitcity.shop.dto.BuyRequest;
import io.github.xpakx.habitcity.shop.dto.ItemResponse;
import io.github.xpakx.habitcity.shop.dto.ShopResponse;

import java.util.List;

public interface ShopService {
    void refreshShops();
    ItemResponse buy(BuyRequest request, Long shopEntryId, Long userId);
    ShopResponse getShop(Long shopId, Long userId);
    void addUserShop(AccountEvent event);
    List<Shop> getShops(Long userId);
}

File: frontend/src/app/shop/shop.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { JwtService } from '../common/jwt-service';
import { BuyRequest } from './dto/buy-request';
import { Shop } from './dto/shop';
import { ShopResponse } from './dto/shop-response';

@Injectable({
  providedIn: 'root'
})
export class ShopService extends JwtService {
  private apiServerUrl = environment.apiServerUrl;

  constructor(private http: HttpClient) {
    super();
  }

  public getShop(shopId: number): Observable<ShopResponse> {
    return this.http.get<ShopResponse>(`${this.apiServerUrl}/shop/${shopId}`, { headers: this.getHeaders() });
  }

  public buy(request: BuyRequest, entryId: number): Observable<ShopResponse> {
    return this.http.post<ShopResponse>(`${this.apiServerUrl}/shop/item/${entryId}`, request, { headers: this.getHeaders() });
  }

  public getShops(): Observable<Shop[]> {
    return this.http.get<Shop[]>(`${this.apiServerUrl}/shop/all`, { headers: this.getHeaders() });
  }

}
